repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -402,7 +400,12 @@ private void traverseEdgesOnDemand(AtlasVertex datasetVertex, boolean isInput, i
                         continue;
                     }
 
-                    if (incrementAndCheckIfRelationsLimitReached(outgoingEdge, isInput, atlasLineageOnDemandContext, ret, depth, baseGuid, direction)) {
+                    // Check for self-cyclic relation on base entity
+                    String datasetGuid = AtlasGraphUtilsV2.getIdFromVertex(datasetVertex);
+                    String entityGuid = AtlasGraphUtilsV2.getIdFromVertex(entityVertex);
+                    boolean baseEntityHasSelfCyclicLineage = datasetGuid.equals(baseGuid) && datasetGuid.equals(entityGuid);
+
+                    if (incrementAndCheckIfRelationsLimitReached(outgoingEdge, isInput, atlasLineageOnDemandContext, ret, depth, baseEntityHasSelfCyclicLineage)) {
```

Comment:
but again, we still skipping output increment for the base entity while processing the first edge.

Reply:
meaning that, we do not increment the output count when we start processing self-cyclic relation. which we should because we want to still process the relation even though it was processed on the INPUT side. 

the only way to know if it is self-cyclic or not while processing on the first edge (from baseGuid to process), is to check the relations.